pkg/nats/publisher.go

```diff
@@ -1,91 +1,135 @@
+type Publisher struct {
+	conn             Connection
+	config           PublisherPublishConfig
+	logger           watermill.LoggerAdapter
+	topicInterpreter *topicInterpreter
 }
 
-// NewStreamingPublisher creates a new StreamingPublisher.
-//
-// When using custom NATS hostname, you should pass it by options StreamingPublisherConfig.StanOptions:
-//		// ...
-//		StanOptions: []stan.Option{
-//			stan.NatsURL("nats://your-nats-hostname:4222"),
-//		}
-//		// ...
-func NewStreamingPublisher(config StreamingPublisherConfig, logger watermill.LoggerAdapter) (*StreamingPublisher, error) {
+// NewPublisher creates a new Publisher.
+func NewPublisher(config PublisherConfig, logger watermill.LoggerAdapter) (*Publisher, error) {
+	config.setDefaults()
+
 	if err := config.Validate(); err != nil {
 		return nil, err
 	}
 
-	conn, err := stan.Connect(config.ClusterID, config.ClientID, config.StanOptions...)
+	conn, err := nats.Connect(config.URL, config.NatsOptions...)
 	if err != nil {
-		return nil, errors.Wrap(err, "cannot connect to nats")
+		return nil, errors.Wrap(err, "cannot connect to nats-core")
 	}
 
-	return NewStreamingPublisherWithStanConn(conn, config.GetStreamingPublisherPublishConfig(), logger)
+	return NewPublisherWithNatsConn(conn, config.GetPublisherPublishConfig(), logger)
 }
 
-func NewStreamingPublisherWithStanConn(conn stan.Conn, config StreamingPublisherPublishConfig, logger watermill.LoggerAdapter) (*StreamingPublisher, error) {
+// NewPublisherWithNatsConn creates a new Publisher with the provided nats-core connection.
+func NewPublisherWithNatsConn(conn *nats.Conn, config PublisherPublishConfig, logger watermill.LoggerAdapter) (*Publisher, error) {
 	if logger == nil {
 		logger = watermill.NopLogger{}
 	}
 
-	return &StreamingPublisher{
-		conn:   conn,
-		config: config,
-		logger: logger,
+	var connection Connection = conn
+	var interpreter *topicInterpreter
+
+	if config.JetStream.Enabled {
+		js, err := conn.JetStream(config.JetStream.ConnectOptions...)
+
+		connection = &jsConnection{conn, js, config.JetStream}
+
+		if err != nil {
+			return nil, err
+		}
+
+		interpreter = newTopicInterpreter(js, config.SubjectCalculator)
+	}
+
+	return &Publisher{
+		conn:             connection,
+		config:           config,
+		logger:           logger,
+		topicInterpreter: interpreter,
 	}, nil
 }
 
 // Publish publishes message to NATS.
 //
-// Publish will not return until an ack has been received from NATS Streaming.
+// Publish will not return until an ack has been received from JetStream.
 // When one of messages delivery fails - function is interrupted.
-func (p StreamingPublisher) Publish(topic string, messages ...*message.Message) error {
+func (p *Publisher) Publish(topic string, messages ...*message.Message) error {
+	// TODO: should we auto provision on publish?  Need durable on publish options...
+	if p.config.JetStream.Enabled && p.config.JetStream.AutoProvision {
```

Comment:
do you think that it may be useful to cache it, like in GCP Pub/Sub? https://github.com/ThreeDotsLabs/watermill-googlecloud/blob/master/pkg/googlecloud/publisher.go#L193

Reply:
Good idea I am not certain we should keep this but minimizing chatter to the broker if publishers do use it makes a ton of sense.